src/stirling/source_connectors/socket_tracer/protocols/mongodb/decode.cc

```diff
@@ -0,0 +1,175 @@
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <libbson-1.0/bson.h>
+
+#include <rapidjson/document.h>
+#include <string>
+
+#include "src/common/base/base.h"
+#include "src/common/base/utils.h"
+#include "src/stirling/source_connectors/socket_tracer/protocols/mongodb/decode.h"
+#include "src/stirling/source_connectors/socket_tracer/protocols/mongodb/types.h"
+#include "src/stirling/utils/binary_decoder.h"
+
+namespace px {
+namespace stirling {
+namespace protocols {
+namespace mongodb {
+
+ParseState ProcessOpMsg(BinaryDecoder* decoder, Frame* frame) {
+  PX_ASSIGN_OR(uint32_t flag_bits, decoder->ExtractLEInt<uint32_t>(), return ParseState::kInvalid);
+
+  // Find relavent flag bit information and ensure remaining bits are not set.
+  // Bits 0-15 are required and bits 16-31 are optional.
+  for (int i = 0; i <= 16; i++) {
+    if (i == 0 && (flag_bits >> i) & 1) {
+      frame->checksum_present = true;
+    } else if (i == 1 && (flag_bits >> i) & 1) {
+      frame->more_to_come = true;
+    } else if (i == 16 && (flag_bits >> i) & 1) {
+      frame->exhaust_allowed = true;
+    } else if (flag_bits >> i) {
+      return ParseState::kInvalid;
+    }
+  }
+
+  // Determine the number of checksum bytes in the buffer.
+  uint8_t checksum_bytes = (frame->checksum_present) ? 4 : 0;
+
+  // Get the section(s) data from the buffer.
+  while (decoder->BufSize() > checksum_bytes) {
+    mongodb::Section section;
+    PX_ASSIGN_OR(section.kind, decoder->ExtractLEInt<uint8_t>(), return ParseState::kInvalid);
+    // Length of the section still remaining in the buffer.
+    int32_t remaining_section_length = 0;
+
+    if (section.kind == 0) {
+      section.length = utils::LEndianBytesToInt<int32_t, 4>(decoder->Buf());
+      if (section.length < kSectionLengthSize) {
+        return ParseState::kInvalid;
+      }
+      remaining_section_length = section.length;
+
+    } else if (section.kind == 1) {
+      PX_ASSIGN_OR(section.length, decoder->ExtractLEInt<uint32_t>(), return ParseState::kInvalid);
+      if (section.length < kSectionLengthSize) {
+        return ParseState::kInvalid;
+      }
+
+      // Get the sequence identifier (command argument).
+      PX_ASSIGN_OR(std::string_view seq_identifier, decoder->ExtractStringUntil('\0'),
+                   return ParseState::kInvalid);
+      // Make sure the sequence identifier is a valid OP_MSG kind 1 command argument.
+      if (seq_identifier != "documents" && seq_identifier != "updates" &&
+          seq_identifier != "deletes") {
+        return ParseState::kInvalid;
+      }
+      remaining_section_length = section.length - kSectionLengthSize - seq_identifier.length() - 1;
+
+    } else {
+      return ParseState::kInvalid;
+    }
+
+    // Extract the document(s) from the section and convert it from type BSON to a JSON string.
+    auto parse_until = decoder->BufSize() - remaining_section_length;
+    while (decoder->BufSize() > parse_until) {
```

Comment:
Is this loop necessary? I know the spec says that zero or more BSON objects will be provided. Reading the `bson_new_from_data` function docs, it appears they refer to this as a document and not as an object. It makes me think that the library _might_ be handling the object looping for us.

<img width="801" alt="Screenshot 2023-09-22 at 2 35 44 PM" src="https://github.com/pixie-io/pixie/assets/5855593/66165c54-e8b8-4bc6-b3f8-864689e46279">

Reply:
I don't think it is necessary to have it. 

However in the case that there are multiple documents (for kind 1 sections only) the no while loop implementation would successfully parse the n number of documents and store them in one document in the `Section` data structure. Having the loop though would let us parse each document individually and store them separately in Section's document vector. I can see why the code might be a unpleasant when handling this scenario. 

We would still need to potentially truncate the data at the time of making the table so maybe it's not that necessary to keep all the docs separated?

Reply:
I'm not questioning if we should parse the entire payload (I believe we should). It's just that there seems to be a discrepancy regarding the difference between a BSON object and a BSON document. I'm asking if we are confident that they are one in the same. Because my reading of the docs seems like a single BSON document could contain multiple BSON objects.

Reply:
I believe they are referring to the same data. The github docs and the wire protocol docs refer to the same portion of the section body as documents and objects. 

https://github.com/mongodb/specifications/blob/master/source/message/OP_MSG.rst#sections:~:text=0%20or%20more%20BSON%20documents.%20Each%20BSON%20document%20cannot%20be%20larger%20than%20maxBSONObjectSize.

 https://www.mongodb.com/docs/v5.2/reference/mongodb-wire-protocol/#:~:text=Zero%20or%20more%20BSON%20objects